File: lib/ts/src/internal/binary/ClientBinaryEncoder.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { ClientBinaryStrategy } from './ClientBinaryStrategy.js';
import { clientBinaryEncode } from '../../internal/binary/ClientBinaryEncode.js';
import { clientBinaryDecode } from '../../internal/binary/ClientBinaryDecode.js';
import { BinaryEncoder } from './BinaryEncoder.js';
import { BinaryEncoding } from './BinaryEncoding.js';
import { BinaryEncodingCache } from './BinaryEncodingCache.js';

export class ClientBinaryEncoder implements BinaryEncoder {
    private readonly binaryEncodingCache: BinaryEncodingCache;
    private readonly binaryChecksumStrategy: ClientBinaryStrategy;

    constructor(binaryEncodingCache: BinaryEncodingCache) {
        this.binaryEncodingCache = binaryEncodingCache;
        this.binaryChecksumStrategy = new ClientBinaryStrategy(binaryEncodingCache);
    }

    encode(message: any[]): any[] {
        return clientBinaryEncode(message, this.binaryEncodingCache, this.binaryChecksumStrategy);
    }

    decode(message: any[]): any[] {
        return clientBinaryDecode(message, this.binaryEncodingCache, this.binaryChecksumStrategy);
    }
}

File: lib/go/internal/binary/ClientBinaryEncoder.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package binary

// ClientBinaryEncoder implements BinaryEncoder for Telepact client messages.
type ClientBinaryEncoder struct {
	cache    BinaryEncodingCache
	strategy *ClientBinaryStrategy
}

// NewClientBinaryEncoder constructs a ClientBinaryEncoder with the provided cache.
func NewClientBinaryEncoder(cache BinaryEncodingCache) *ClientBinaryEncoder {
	return &ClientBinaryEncoder{
		cache:    cache,
		strategy: NewClientBinaryStrategy(cache),
	}
}

// Encode encodes a pseudo-JSON client message body into its binary representation.
func (e *ClientBinaryEncoder) Encode(message []any) ([]any, error) {
	return ClientBinaryEncode(message, e.cache, e.strategy)
}

// Decode decodes a binary client message body back into pseudo-JSON.
func (e *ClientBinaryEncoder) Decode(message []any) ([]any, error) {
	return ClientBinaryDecode(message, e.cache, e.strategy)
}
